torch/csrc/autograd/python_variable_indexing.cpp

```diff
@@ -247,6 +247,76 @@ static PyObject* applyBoolGetitem(const Variable& self, bool index) {
   }
 }
 
+static bool canDispatchToLegacyIndexing(const Variable& self, const variable_list& vars) {
+  // mask indexing
+  if (vars.size() == 1 && vars[0].type().scalarType() == ScalarType::Byte && vars[0].is_same_size(self)) {
+    return true;
+  }
+
+  // single tensor indexing
+  int num_defined_variables = 0;
+  for (auto& variable : vars) {
+    auto is_defined = variable.defined();
+    num_defined_variables += is_defined;
+    if (is_defined && (variable.dim() != 1 || variable.type().scalarType() != ScalarType::Long || variable.numel() == 0)) {
+      num_defined_variables = -1;
+      break;
+    }
+  }
+  if (num_defined_variables == 1) {
+    return true;
+  }
+  // advanced indexing
+  return false;
+}
+
+static Variable dispatch_legacy_index(const Variable& self, const variable_list& vars) {
+  if (vars.size() == 1 && vars[0].type().scalarType() == ScalarType::Byte && vars[0].is_same_size(self)) {
+    auto mask = vars[0];
+    if (self.type().backend() != mask.type().backend())
+      mask = mask.toBackend(self.type().backend());
+    return self.masked_select(mask);
+  }
+
+  int64_t index_dim = -1;
+  for (size_t i = 0; i < vars.size(); i++) {
+    if (vars[i].defined()) {
+      index_dim = i;
+      break;
+    }
+  }
+  auto index = vars[index_dim];
+  if (self.type().backend() != index.type().backend())
+    index = index.toBackend(self.type().backend());
+  return self.index_select(index_dim, index);
+}
+
+static Variable dispatch_legacy_index_put(Variable& self, const variable_list& vars, const Variable& value) {
+  auto value_ = value;
+  if (self.type().backend() != value_.type().backend())
+    value_ = value_.toBackend(self.type().backend());
```

Comment:
this is inconsistent with the non-legacy case, e.g. today:
>>> a[idx]=torch.tensor(8).cuda()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: Expected object of type torch.cuda.FloatTensor but found type torch.cuda.LongTensor for argument #3 'source'

Reply:
Good catch!